### PR TITLE
Add vector label support

### PR DESCRIPTION
**Pull request by loicgasser**
This PR adds:
* labels support to Vector Layers (points, lines and polygons) via JSON config files.
* additional tests
* a specification document to configure vector layers

I provide a couple of test links/show cases tomorrow.

### CODE REVIEW

**Comment by loicgasser** [comment]
(Test Link labels + resolutions)[https://mf-geoadmin3.dev.bgdi.ch/gal_meteo/?layers=ch.bafu.hydroweb-warnkarte_national,ch.swissmetnet.temperature&topic=ech&lang=en&bgLayer=ch.swisstopo.swissimage&X=184250.00&Y=659250.00&zoom=2]
(Related config file)[https://github.com/geoadmin/mf-chsdi3/blob/79de1baee420edf01e6eb488201f86c3357f13a6/chsdi/static/vectorStyles/ch.swissmetnet.temperature.json]
Ready for review.

**Comment by AFoletti** [comment]
I like it. The doc too!

**Comment by gjn** [comment]
Awsome!

Doc could be in official API3 documentation. I could imagine people integrating/importing their on GeoJSON layers using their custom styles.

In general, styling will/is also an issue for vector tiles - no clear definition yet how to specify those. But it would be good to be able to use the same styling definitions for any vector layer and vector tiles. Might need deeper discussion.

@oterral @davidoesch Thougths on this?

Small tidbit regarding this data set...
[Generalisation...](https://mf-geoadmin3.dev.bgdi.ch/gal_meteo/?layers=ch.bafu.hydroweb-warnkarte_national,ch.swissmetnet.temperature&topic=ech&lang=en&bgLayer=ch.swisstopo.swissimage&X=146061.73&Y=552621.73&zoom=8&layers_visibility=true,false)

**Comment by loicgasser** [comment]
>>> But it would be good to be able to use the same styling definitions for any vector layer and vector tiles. Might need deeper discussion.

Yes, that was my intent. It is designed to work with any OL3 vector layer.

I'd like to work on the writing part and eventually create a small editor.

**Comment by gjn** [comment]
@oterral Please review and merge.

File: src/components/StylesFromLiteralsService.js [hunk 1]
```diff
@@ -59,32 +61,65 @@ goog.provide('ga_stylesfromliterals_service');
...
+            olGeometry instanceof ol.geom.MultiPolygon) {
+          return 'polygon';
+        }
+      }
 
+      function getLabelProperty(value) {
+        if (value) {
+          return value.property !== undefined ? value.property : null;
+        }
+        return null;
```
Comment (oterral) [inline src/components/StylesFromLiteralsService.js]: Any reason to return null value ? 

If undefined ois also correct you can simplify this line to : 
```
if (value) {
  return value.property;
}
```

File: src/components/StylesFromLiteralsService.js [hunk 2]
```diff
@@ -59,32 +61,65 @@ goog.provide('ga_stylesfromliterals_service');
...
+      }
+
+      function getMinResolution(value) {
+        return value.minResolution !== undefined ? value.minResolution : 0;
```
Comment (oterral) [inline src/components/StylesFromLiteralsService.js]: return value.minResolution || 0

File: src/components/StylesFromLiteralsService.js [hunk 3]
```diff
@@ -59,32 +61,65 @@ goog.provide('ga_stylesfromliterals_service');
...
+      }
+
+      function getMaxResolution(value) {
+        return value.maxResolution !== undefined ? value.maxResolution :
```
Comment (oterral) [inline src/components/StylesFromLiteralsService.js]: return value.maxResolution || 0

File: src/components/StylesFromLiteralsService.js [hunk 4]
```diff
@@ -93,113 +128,131 @@ goog.provide('ga_stylesfromliterals_service');
...
-            }
+      olStyleForPropertyValue.prototype.findOlStyleInRange_ = function(value,
+          geomType) {
+        var olStyle, range;
+        for (range in this.styles[geomType]) {
+          range = range.split(',');
+          if (value >= parseFloat(range[0]) &&
+              value < parseFloat(range[1])) {
+            var style = this.styles[geomType][range];
+            olStyle = style;
```
Comment (oterral) [inline src/components/StylesFromLiteralsService.js]: olStyle = this.styles[geomType][range];

File: src/components/StylesFromLiteralsService.js [hunk 5]
```diff
@@ -93,113 +128,131 @@ goog.provide('ga_stylesfromliterals_service');
...
-      }
+          return this.singleStyle.olStyle;
+        } else if (this.type === 'unique') {
+          var value;
+          var properties = feature.getProperties();
+          if (this.key !== undefined) {
+            value = properties[this.key];
+          } else {
+            value = this.defaultVal;
+          }
```
Comment (oterral) [inline src/components/StylesFromLiteralsService.js]: `var value = properties[this.key] || this.defaultVal`

**Comment by oterral** [comment]
Lgtm , only some minors code simplification

**Comment by oterral** [comment]
Really nice feature !!!!

File: src/components/StylesFromLiteralsService.js (hunk 1 above)
Comment (loicgasser) [inline src/components/StylesFromLiteralsService.js]: yes makes sense thanks

**Comment by loicgasser** [comment]
Thank you for the review! I'll address your points

**Comment by loicgasser** [comment]
Comments addressed and rebased

**Pull request merged**